lnst/Recipes/ENRT/BaseTunnelRecipe.py

```diff
@@ -1,8 +1,20 @@
+from typing import Optional, Union
+
+from lnst.Devices import Device, RemoteDevice
 from lnst.Recipes.ENRT.BaremetalEnrtRecipe import BaremetalEnrtRecipe
 from lnst.RecipeCommon.PacketAssert import PacketAssertTestAndEvaluate
 from lnst.Recipes.ENRT.BaseEnrtRecipe import EnrtConfiguration
 
 
+AnyDevice = Union[Device, RemoteDevice]
```

Comment:
When creating software devices (such as `GreTunnelRecipe`'s `GreDevice`), they inherit from `Device`, not from `RemoteDevice`. These devices are then put into `tunnel_devices`. What should I do in this case?

Reply:
you mean this? https://github.com/LNST-project/lnst/blob/master/lnst/Recipes/ENRT/GeneveLwtTunnelRecipe.py#L134

right well... there's a bit of python forbidden magic at play here. The way that device classes work in lnst is that the constructs:

```python
Host.new_dev_var = SomeDevice()
```

actually creates a `RemoteDevice(SomeDevice)` instance that is assigned to the `Host` object, and on the assignment call (so `setattr`) we detect a RemoteDevice assignment and translate it into a configuration command sent to the agent where the actual `SomeDevice` instance is created.

so the order of operations is:
1. import SomeDevice from `lnst.Devices` instead of `lnst.Devices.SomeDevice` e.g. - https://github.com/LNST-project/lnst/blob/master/lnst/Recipes/ENRT/GeneveLwtTunnelRecipe.py#L10 imports GeneveDevice from `lnst.Device`
2. https://github.com/LNST-project/lnst/blob/master/lnst/Devices/__init__.py#L48 `lnst.Device.GeneveDevice` is actually wrapped as a `RemoteDevice(Geneve) here
3. https://github.com/LNST-project/lnst/blob/master/lnst/Recipes/ENRT/GeneveLwtTunnelRecipe.py#L134 creates a `RemoteDevice(GeneveDevice)` instance
4. the assignment goes to https://github.com/LNST-project/lnst/blob/master/lnst/Controller/Namespace.py#L157 where we detect a RemoteDevice and send a create device command to the agent.

so on the controller, you're never actually using the Device class instances.... YOU MAY however use them for instance checks:
```python
from lnst.Device import RemoteGeneveDevice
from lnst.Device.GeneveDevice import GeneveDevice

m1.geneve_device = RemoteGeneveDevice(...)

isinstance(m1.geneve_device, GeneveDevice) == True
isinstance(m1.geneve_device, RemoteGeneveDevice) == True
```

Reply:
Right... this is confusing and very type unsafe resulting in a lot of errors like
```
Missing positional argument "ifmanager" in call to "GeneveDevice"
```
I wish this was handled better but it seems we're stuck like this, at least for some time.

I'll use only `RemoteDevice` typehints then.

Reply:
> Missing positional argument "ifmanager" in call to "GeneveDevice"

this can only happen if the type checker is using the incorrect thing... we're not actually importing `GeneveDevice` in the recipe... we're importing the decorated `RemoteDevice` and that doesn't take the ifmanager parameter. So if the type checker correctly used the right objects this error shouldn't appear.

But that's a separate discussion, let's have that at some coffee time or something instead of here...

Reply:
true, but I don't think type-checkers can enforce this when we're editing `globals()` using strings. Anyway, I removed the Union and used `RemoteDevice` instead.